client/js/s3/request-signer.js

```diff
@@ -302,6 +307,32 @@ qq.s3.RequestSigner = function(o) {
             errorMessage = "Received an empty or invalid response from the server!";
         }
 
+        response = options.onSigningRequestComplete(response, !isError, xhrOrXdr, tempCredentials);
+
+        if (tempCredentials) {
+            if (Object.keys) {
+                Object.keys(tempCredentials).length === 0 && (tempCredentials = null);
+            }
+            else {
+                /** this is for pre-ECMA 5 compatability */
+                for (i in tempCredentials) {
+                    if (tempCredentials.hasOwnProperty(i)) {
+                        tempCredentials = null;
+                        break;
+                    }
+                }
+            }
+        }
+
+        if (tempCredentials) {
+            options.log("Temporary Amazon credentials found after onSigningRequestComplete event callback");
+            credentials = tempCredentials;
+        } else {
+            credentials = credentialsProvider.get();
+        }
+
+        isError = isError || (!response) && Boolean(errorMessage = "onSigningRequestComplete event callback returned false");
```

Comment:
This is pretty dense hard to follow. Also, I'm not sure about determining if an error exists based on text in the error message.